core/src/main/scala/akka/persistence/r2dbc/journal/JournalDao.scala

```diff
@@ -242,58 +249,99 @@ private[r2dbc] class JournalDao(journalSettings: R2dbcSettings, connectionFactor
+              deletedRows.toString, //toString for fix scala 2.12 compilation error
+              persistenceId,
+              from.toString,
+              to.toString)
+          })(ExecutionContexts.parasitic)
+    }
 
-        val entityType = PersistenceId.extractEntityType(persistenceId)
-        val slice = persistenceExt.sliceForPersistenceId(persistenceId)
-
-        val deleteMarkerSeqNrFut =
-          if (toSequenceNr == Long.MaxValue)
-            readHighestSequenceNr(persistenceId, 0L)
-          else
-            Future.successful(toSequenceNr)
-
-        deleteMarkerSeqNrFut.flatMap { deleteMarkerSeqNr =>
-          def bindDeleteMarker(stmt: Statement): Statement = {
-            stmt
-              .bind(0, slice)
-              .bind(1, entityType)
-              .bind(2, persistenceId)
-              .bind(3, deleteMarkerSeqNr)
-              .bind(4, "")
-              .bind(5, "")
-              .bind(6, 0)
-              .bind(7, "")
-              .bind(8, Array.emptyByteArray)
-              .bind(9, true)
-          }
-
-          r2dbcExecutor
-            .update(s"delete [$persistenceId]") { connection =>
-              Vector(
-                connection
-                  .createStatement(deleteEventsSql)
-                  .bind(0, persistenceId)
-                  .bind(1, toSequenceNr),
-                bindDeleteMarker(connection.createStatement(insertDeleteMarkerSql)))
-            }
-            .map(_.head)
-        }
+    val batchSize = journalSettings.cleanupSettings.eventsJournalDeleteBatchSize
+
+    def deleteInBatches(from: Long, maxTo: Long): Future[Unit] = {
+      if (from + batchSize > maxTo) {
+        deleteBatch(from, maxTo)
+      } else {
+        val to = from + batchSize - 1
+        deleteBatch(from, to).flatMap(_ => deleteInBatches(to + 1, maxTo))
       }
+    }
 
-    if (log.isDebugEnabled)
-      result.foreach(deletedRows => log.debug("Deleted [{}] events for persistenceId [{}]", deletedRows, persistenceId))
+    def insertDeleteMarker(deleteMarkerSeqNr: Long): Future[Long] = {
+      val entityType = PersistenceId.extractEntityType(persistenceId)
+      val slice = persistenceExt.sliceForPersistenceId(persistenceId)
+      r2dbcExecutor
+        .updateOne(s"insert delete marker for [$persistenceId], seq num [$deleteMarkerSeqNr]") { connection =>
+          connection
+            .createStatement(insertDeleteMarkerSql)
+            .bind(0, slice)
+            .bind(1, entityType)
+            .bind(2, persistenceId)
+            .bind(3, deleteMarkerSeqNr)
+            .bind(4, "")
+            .bind(5, "")
+            .bind(6, 0)
+            .bind(7, "")
+            .bind(8, Array.emptyByteArray)
+            .bind(9, true)
+        }
+    }
 
-    result.map(_ => ())(ExecutionContexts.parasitic)
+    if (resetSequenceNumber) {
+      for {
+        fromSeqNr <- readLowestSequenceNr(persistenceId)
+        toSeqNr <- highestSeqNr(persistenceId, toSequenceNr)
```

Comment:
If we read highest first then we can decide if the readLowestSequenceNr is needed or not. If highest is less than the batch size we know that there is no need for batching and lowest is not needed (or can be set to 1L).

Reply:
fixed